File: spendy_backend/src/main/java/com/rohitcodes/expense_tracker/service/CategoryService.java
package com.rohitcodes.expense_tracker.service;

import com.rohitcodes.expense_tracker.entity.CategoryType;
import com.rohitcodes.expense_tracker.dto.category.CategoryRequest;
import com.rohitcodes.expense_tracker.dto.category.CategoryResponse;

import java.util.List;

public interface CategoryService {

    CategoryResponse createCategory(Long userId, CategoryRequest request);

    List<CategoryResponse> getCategories(Long userId, CategoryType type);

    CategoryResponse getCategoryById(Long userId, Long categoryId);

    CategoryResponse updateCategory(Long userId, Long categoryId, CategoryRequest request);

    void deleteCategory(Long userId, Long categoryId);
}

File: spendy-mobileFrontend/src/services/categoryService.ts
import { api } from "@/src/services/api";
import type { Category, CategoryPayload, CategoryType } from "@/src/types/category";

export const categoryService = {
  async getCategories(userId: number, type?: CategoryType) {
    const response = await api.get<Category[]>(`/users/${userId}/categories`, {
      params: type ? { type } : undefined
    });
    return response.data;
  },

  async getCategoryById(userId: number, categoryId: number) {
    const response = await api.get<Category>(`/users/${userId}/categories/${categoryId}`);
    return response.data;
  },

  async createCategory(userId: number, payload: CategoryPayload) {
    const response = await api.post<Category>(`/users/${userId}/categories`, payload);
    return response.data;
  },

  async updateCategory(userId: number, categoryId: number, payload: CategoryPayload) {
    const response = await api.put<Category>(`/users/${userId}/categories/${categoryId}`, payload);
    return response.data;
  },

  async deleteCategory(userId: number, categoryId: number) {
    await api.delete(`/users/${userId}/categories/${categoryId}`);
    return categoryId;
  }
};
